### DISCUSSION TITLE
Reading a C char array as a Go string

### DISCUSSION THREAD

**Discussion by tamilmani1989** [cilium/ebpf#648] [Q&A] [answered]
Hi,

My ebpf program gets process name which is retrieved as char array and i use bpf2go to generate types in golang.  The golang type converting it as int8[16] and how would i convert that to string in golang?

```
struct verdict
{
	struct iptuple flow;
	__u32 pid;
	char comm[TASK_COMM_LEN];
	int status;
};
bpf_get_current_comm(v.comm, sizeof(v.comm));
```

Generated type in Go code:

```
type bpfVerdict struct {
	Flow   bpfIptuple
	Pid    uint32
	Comm   [16]int8
	Status int32
}
```

**Comment by mmat11** [comment, marked as answer]
You could define your struct in C code as:

```
struct verdict {
    ..
    __u8 comm[TASK_COMM_LEN];
    ..
};
```

which will generate `Comm [16]uint8`.

Edit: as Lorenz pointed out below, this field will likely contain a nul-terminated C string sent by the kernel, which can be parsed using https://pkg.go.dev/golang.org/x/sys/unix#ByteSliceToString. See https://go.dev/play/p/16gaV3H2yM-.

**Reply by lmb** [reply]
You'll probably want to use https://pkg.go.dev/golang.org/x/sys/unix#ByteSliceToString to deal with \x00.

**Reply by lmb** [reply]
Some more background on why char turns into int8: https://stackoverflow.com/a/2054941 Seems like you could also try -funsigned-char but then who knows what implications that has.

**Reply by tamilmani1989** [reply]
Thank you guys! It worked :)
